test/ProductConstructionService.ScenarioTests/CodeFlowScenarioTestBase.cs

```diff
@@ -0,0 +1,155 @@
+
+        try
+        {
+            var versionDetailsFile = files.FirstOrDefault(file => file.FileName == "eng/Version.Details.xml");
+            versionDetailsFile.Should().NotBeNull();
+            versionDetailsFile!.Patch.Should().Contain(GetExpectedCodeFlowDependencyVersionEntry(sourceRepoName, commitSha));
+
+            // Verify new files are in the PR
+            foreach (var testFile in testFiles)
+            {
+                var newFile = files.FirstOrDefault(file => file.FileName == $"{testFile}");
+                newFile.Should().NotBeNull();
+                newFile!.Patch.Should().Be(testFilePatches[testFile]);
+            }
+        }
+        finally
+        {
+            // close the PR
+            await GitHubApi.PullRequest.Update(
+                _parameters.GitHubTestOrg,
+                targetRepoName,
+                pullRequest.Number,
+                new Octokit.PullRequestUpdate
+                {
+                    State = Octokit.ItemState.Closed
+                });
+        }
+    }
+
+    protected async Task<AsyncDisposableValue<string>> CreateSourceEnabledSubscriptionAsync(
+        string sourceChannelName,
+        string sourceRepo,
+        string targetRepo,
+        string targetBranch,
+        string updateFrequency,
+        string sourceOrg = "dotnet",
+        List<string>? additionalOptions = null,
+        bool sourceIsAzDo = false,
+        bool targetIsAzDo = false,
+        bool trigger = false,
+        string? sourceDirectory = null,
+        string? targetDirectory = null)
+    {
+        if (!string.IsNullOrEmpty(sourceDirectory) && !string.IsNullOrEmpty(targetDirectory))
+        {
+            throw new ScenarioTestException("When creating a source enabled subscription, either provide a 'source-directory' or a 'target-directory', never both");
+        }
+
+        if (!string.IsNullOrEmpty(sourceDirectory))
+        {
+            return await CreateSubscriptionAsync(
+                sourceChannelName,
+                sourceRepo,
+                targetRepo,
+                targetBranch,
+                updateFrequency,
+                sourceOrg,
+                [   .. additionalOptions,
+                    "--source-enabled", "true",
+                    "--source-directory", sourceDirectory ],
+                sourceIsAzDo,
+                targetIsAzDo,
+                trigger);
+        }
+
+        if (!string.IsNullOrEmpty(targetDirectory))
+        {
+            return await CreateSubscriptionAsync(
+                sourceChannelName,
+                sourceRepo,
+                targetRepo,
+                targetBranch,
+                updateFrequency,
+                sourceOrg,
+                [   .. additionalOptions,
+                    "--source-enabled", "true",
+                    "--target-directory", targetDirectory ],
+                sourceIsAzDo,
+                targetIsAzDo,
+                trigger);
+        }
```

Comment:
These block are the same except the argument name? Maybe we can just do

- if both set, exception
- if none set, exception
- `return CreateSubscriptionAsync(... [ (.. ? "--source-directory" : "--target-directory") ] ...)`

But overall, if you provide a method which can accept invalid mix of optional arguments, it's often better to just have two of them `CreateForwardFlowSubscriptionAsync` and `CreateBackwardFlowSubscriptionAsync` and have those call with the arguments set the right way.

Reply:
Yeah, that's good advice, thanks